standup/auth0/settings.py

```diff
@@ -0,0 +1,212 @@
+
+            This is used every time we need to logout a user and redirect them to a
+            signin form.
+            """)
+        ),
+        Option(
+            'AUTH0_PATIENCE_TIMEOUT', default=5, parser=int,
+            help_text=dedent("""\
+            The amount of time in seconds that your app will wait when
+            sending requests to the Auth0 server.
+            """)
+        ),
+        Option(
+            'AUTH0_ID_TOKEN_DOMAINS',
+            help_text=dedent("""\
+            list of strings: The domains that require an ``id_token`` for your app.
+
+            For Mozilla sites, this is something like::
+
+                ['mozilla.com', 'mozillafoundation.org', 'mozilla-japan.org']
+
+            If someone logs into Auth0 using an account that has a verified email address
+            with one of those domains, then they'll be required to have logged in using
+            an Oauth2 Auth0 provider like the Mozilla LDAP option. If they didn't use
+            such a provider, they're immediately logged out and told to use an
+            appropriate provider.
+            """)
+        ),
+        Option(
+            'AUTH0_ID_TOKEN_EXPIRY', default=900, parser=int,
+            help_text=dedent("""\
+            Users who used an Oauth2 Auth0 provider and have a verified
+            email address in ``AUTH0_ID_TOKEN_DOMAINS`` have their ``id_token`` renewed
+            every ``AUTH0_ID_TOKEN_EXPIRY`` seconds.
+
+            If the ``id_token`` fails renewal, the user is immediately logged out.
+            """)
+        )
+    ]
+
+    def __init__(self):
+        self._option_map = OrderedDict([
+            (opt.key, opt) for opt in self._all_options
+        ])
+
+    def get(self, key):
+        opt = self._option_map[key]
+        val = getattr(settings, opt.key, NO_VALUE)
+        if val is not NO_VALUE:
+            try:
+                return opt.parser(val)
+            except Exception:
+                msg = '"%s": "%s" kicked up error when parsing.' % (opt.key, val)
+                if opt.help_text:
+                    msg = '%s\n\n%s' % (msg, opt.help_text)
+                raise ConfigurationError(msg)
+
+        if opt.default is not NO_VALUE:
+            try:
+                return opt.parser(opt.default)
+            except Exception:
+                msg = '"%s": "%s" (default) kicked up error when parsing.' % (opt.key, val)
+                if opt.help_text:
+                    msg = '%s\n\n%s' % (msg, opt.help_text)
+                raise ConfigurationError(msg)
+
+        msg = '"%s" must be configured in settings.' % opt.key
+        if opt.help_text:
+            msg = '%s\n\n%s' % (msg, opt.help_text)
+        raise ConfigurationError(msg)
+
+    def __getattr__(self, key):
+        if key.startswith('AUTH0'):
+            return self.get(key)
+        super().__getattr__(self, key)
+
+
+app_settings = AppSettings()
```

Comment:
This is my new favorite way to do settings in Django libraries. It looks uninteresting here, but it's trivial to write a script that runs through the options and generates reST doc for Sphinx documentation with index points, formatting, etc. That makes it a lot easier to make sure docs match the code.

Plus it will throw helpful configuration errors. For example:

```
[13/Jan/2017 15:08:42] ERROR django.request: Internal Server Error: /statistics/
Traceback (most recent call last):
  File "/usr/local/lib/python3.5/dist-packages/django/core/handlers/base.py", line 132, in get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/app/standup/status/views.py", line 319, in statistics
    'statsitems': groups,
  File "/usr/local/lib/python3.5/dist-packages/django/shortcuts.py", line 67, in render
    template_name, context, request=request, using=using)
  File "/usr/local/lib/python3.5/dist-packages/django/template/loader.py", line 99, in render_to_string
    return template.render(context, request)
  File "/usr/local/lib/python3.5/dist-packages/django_jinja/backend.py", line 84, in render
    context.update(processor(request))
  File "/app/standup/auth0/context_processors.py", line 40, in auth0
    'auth0configured': is_auth0_configured(),
  File "/app/standup/auth0/context_processors.py", line 9, in is_auth0_configured
    app_settings.AUTH0_CLIENT_ID and
  File "/app/standup/auth0/settings.py", line 215, in __getattr__
    return self.get(key)
  File "/app/standup/auth0/settings.py", line 211, in get
    raise ConfigurationError(msg)
standup.auth0.settings.ConfigurationError: "AUTH0_CLIENT_ID" must be configured in settings.

Your Auth0 client id. This is public-ish.
```